access_control.py

```diff
@@ -0,0 +1,428 @@
+    """
+    print("Processing access control request")
+    granted, failed = _process_access_control_requests("grant_access", remove=False)
+    revoked, failed_2 = _process_access_control_requests("revoke_access", remove=True)
+    return granted, revoked, failed + failed_2
+
+
+def _process_request_for_feedstock(
+    feedstock: str,
+    resource: str,
+    remove: bool,
+    cirun_policy_args: Optional[List[str]] = None,
+) -> None:
+    """
+    Process the access control request for a single feedstock.
+
+    Parameters:
+    feedstock (str): The name of the feedstock.
+    resource (str): The name of the resource for access control.
+    remove (bool): Whether to remove the access control.
+    cirun_policy_args (List[str]): A list of policy arguments for cirun resources.
+    """
+    with tempfile.TemporaryDirectory() as tmp_dir:
+        feedstock_dir = os.path.join(tmp_dir, feedstock)
+        assert GH_ORG
+        clone_cmd = [
+            "git",
+            "clone",
+            "--depth",
+            "1",
+            f"https://github.com/{GH_ORG}/{feedstock}.git",
+            feedstock_dir,
+        ]
+        print("Cloning:", *clone_cmd)
+        subprocess.check_call(clone_cmd)
+
+        owner_info = ["--organization", GH_ORG]
+        token_repo = (
+            'https://x-access-token:${GITHUB_TOKEN}@github.com/'
+            f'{GH_ORG}/feedstock-tokens'
+        )
+
+        register_ci_cmd = [
+            "conda-smithy",
+            "register-ci",
+            "--feedstock_dir",
+            feedstock_dir,
+            "--without-all",
+            "--without-anaconda-token",
+            *owner_info,
+        ]
+        if resource == "travis":
+            register_ci_cmd.append("--with-travis")
+
+        elif resource.startswith("cirun-"):
+            register_ci_cmd.extend(
+                [
+                    "--with-cirun",
+                    "--cirun-resources", resource,
+                ]
+            )
+            if cirun_policy_args:
+                for arg in cirun_policy_args:
+                    register_ci_cmd.extend(("--cirun-policy-args", arg))
+
+            if remove:
+                register_ci_cmd.append("--remove")
+
+        print("Register-CI command:", *register_ci_cmd)
+        subprocess.check_call(register_ci_cmd)
+
+        if not remove:
+            if resource == "travis":
+                with_cmd = "--with-travis"
+            elif resource.startswith("cirun-"):
+                with_cmd = "--with-github-actions"
+
+            print("Generating a new feedstock token")
+            subprocess.check_call(
+                [
+                    'conda', 'smithy', 'generate-feedstock-token',
+                    '--unique-token-per-provider',
+                    '--feedstock_directory', feedstock_dir,
+                    *owner_info,
+                ]
+            )
+
+            print("Register new feedstock token with provider and feedstock-tokens repo.")
```

Comment:
We need to double check that smithy, our tokens repo, and the webservices are properly interacting with the unique tokens. I don't recall how much testing I have done.

Reply:
I checked and it works as intended. (At least for github actions as the provider)

Reply:
Ah just so I am clear. You looked in the tokens repo, the json was formatted correctly, and the tokens on the other CI providers we did not touch still work?

Reply:
Yes to both.

Reply:
great! thank you for making sure all of the details worked!